src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,409 @@
+
+    if (!qmlEngine->incubationController()) {
+        qmlEngine->setIncubationController(m_quickWindow->incubationController());
+    }
+
+    m_quickWindow->setGeometry(0, 0, m_screenInfo.size.width(), m_screenInfo.size.height());
+
+    m_isValid = true;
+    m_waitCondition.wakeAll();
+}
+
+void ControllerRenderingEngine::finish() {
+    m_isValid = false;
+    disconnect(this);
+
+    const auto lock = lockMutex(&s_glMutex);
+
+    if (m_context && m_context->isValid()) {
+        disconnect(m_context.get(),
+                &QOpenGLContext::aboutToBeDestroyed,
+                this,
+                &ControllerRenderingEngine::finish);
+        m_context->makeCurrent(m_offscreenSurface.get());
+        m_renderControl.reset();
+
+        std::shared_ptr<QOffscreenSurface> pOffscreenSurface = std::move(m_offscreenSurface);
+        QMetaObject::invokeMethod(
+                qApp,
+                [pOffscreenSurface] {
+                    pOffscreenSurface->destroy();
+                });
+        m_quickWindow.reset();
+
+        // Free the engine and FBO
+        m_fbo.reset();
+
+        m_context->doneCurrent();
+    }
+    m_context.reset();
+    m_pThread->quit();
+}
+
+void ControllerRenderingEngine::renderFrame() {
+    if (!m_isValid) {
+        DEBUG_ASSERT(!"Trying to render frame on an invalid engine");
+        return;
+    }
+
+    VERIFY_OR_DEBUG_ASSERT(m_offscreenSurface->isValid()) {
+        kLogger.warning() << "OffscreenSurface isn't valid anymore.";
+        finish();
+        return;
+    };
+    VERIFY_OR_DEBUG_ASSERT(m_context->isValid()) {
+        kLogger.warning() << "GLContext isn't valid anymore.";
+        finish();
+        return;
+    };
+
+    auto lock = lockMutex(&s_glMutex);
+
+    VERIFY_OR_DEBUG_ASSERT(m_context->makeCurrent(m_offscreenSurface.get())) {
+        kLogger.warning() << "Couldn't make the GLContext current to the OffscreenSurface.";
+        lock.unlock();
+        finish();
+        return;
+    };
+
+    if (!m_fbo) {
+        VERIFY_OR_DEBUG_ASSERT(QOpenGLFramebufferObject::hasOpenGLFramebufferObjects()) {
+            kLogger.warning() << "OpenGL doesn't support FBO";
+            lock.unlock();
+            finish();
+            return;
+        };
+
+        m_fbo = std::make_unique<QOpenGLFramebufferObject>(
+                m_screenInfo.size, QOpenGLFramebufferObject::CombinedDepthStencil);
+
+        GLenum glError;
+        glError = m_context->functions()->glGetError();
+
+        VERIFY_OR_DEBUG_ASSERT(glError == GL_NO_ERROR) {
+            kLogger.warning() << "GLError: " << glError;
+            lock.unlock();
+            finish();
+            return;
+        };
+
+        VERIFY_OR_DEBUG_ASSERT(m_fbo->isValid()) {
+            kLogger.warning() << "Failed to initialize FBO";
+            lock.unlock();
+            finish();
+            return;
+        };
+
+        m_quickWindow->setGraphicsDevice(QQuickGraphicsDevice::fromOpenGLContext(m_context.get()));
```

Comment:
I do not feel qualified to review new GL code, but it all looks ok 😅

Reply:
Perhaps @m0dB could help with this section? 
For the record, this is the section that is likely to require some refactor in Qt6.6+ if we want support new RHI (Wayland, but also DirectX and Metal, as currently, Windows and Mac will use OpenGL for this, but could likely benefit of better performance with their native platform)